src/libraries/System.AppContext/src/System.AppContext.csproj

```diff
@@ -2,9 +2,9 @@
   <PropertyGroup>
     <AssemblyName>System.AppContext</AssemblyName>
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
-    <TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;$(NetCoreAppCurrent)-Unix</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent)</TargetFrameworks>
   </PropertyGroup>
   <ItemGroup>
-    <ReferenceFromRuntime Include="System.Private.CoreLib" />
+    <Reference Include="System.Runtime" />
```

Comment:
Going through System.Runtime is an extra forwarder at runtime that adds a bit of inefficiency.

Are we getting anything in return for this bit of inefficiency?

Reply:
Yeah, I don't think we want this if this assembly is actually exposing types from S.P.CoreLib.

Reply:
@Jkotas, the type is forward in the ref as well, so the only code that will load System.AppContext at runtime is legacy code compiled against a previous release.  In that case do you still care?

The benefit we're getting from this is a simpler build where we don't need a project reference to CoreLib's implementation.  Previously it would also mean we could eliminate configurations, but we don't need to remove the corelib reference to do that any more based on discussion in https://github.com/dotnet/runtime/issues/34653, so its really just about the ProjectReference to CoreLib.

Reply:
I agree that this will only matters for code compiled against older contracts.

What is the build simplification we get by replacing System.Private.CoreLib reference with System.Runtime reference? The two options look equivalent to me.

Reply:
Simple `References` depend on public API which is satisfied by an up front build of all references assemblies: a quick and shallow build, outside the context of the src project build.  This enables us to flatten the build graph so that most src assemblies become fully parallelizable.  It also allows us to preserve our desire to only reference public API between projects.  It also enables simple solutions that only contain a single library.  There are further benefits to this approach that reduce the time of incremental builds, since most projects only depend on public API of reference assemblies that don't change much.

A `ReferenceFromRuntime` turns into a `ProjectRefeference` to implementation.  Since this is a different assembly universe, the reference becomes viral and means that anything in this space must use `ProjectReferences` to other src projects.  This deepens the build graph: src projects which were previously parallelizable are no longer.  It also complicates the development experience: `ProjectReferences` require the projects to be present in the SLN so iterating on a single library needs to build the entire src graph for that library.  Similarly it increases the cost of incremental builds.  Src projects (and corelib) are more likely to change, so incremental build will rebuild more if you have a deeper project graph.

For these reasons, we have the general rule: Do not reference System.Private.Corelib unless you **need** types only defined in System.Private.Corelib and not public reference assemblies.

In this specific case, its not too problematic, since this library is a leaf (so build graph doesn't really suffer all that much) and no longer contains any types (so development experience isn't important as the only development will be on the project file).  It does still increase the cost of an incremental build that changes CoreLib. 

I'm more concerned with the generalization of your argument, which suggests that *everything* in the shared framework should reference System.Private.CoreLib.  I don't think we should go there for the reasons I called out above.  I don't want to have a bunch of special cases to our rule about referencing CoreLib.  In fact, I was hoping that your effort to push types down into System.Runtime would make our build be able to eliminate nearly all of the cases where we reference CoreLib (effectively making System.Runtime the public contract for CoreLib).

One thing I've debated, is moving all these legacy contracts which are now full-facades into shims, so that we don't need to think about them in the same way as the rest of the src project and can effectively fully-generate their content.  The shims are sequenced after the src build and rarely change so they don't suffer from most of the problems I mentioned above.

Reply:
> I'm more concerned with the generalization of your argument, which suggests that everything in the shared framework should reference System.Private.CoreLib. 

Yes, it would be nice for the shipping shared framework bits to avoid the unnecessary extra hops. I understand why we have opposite preference for the inner loop.

These extra hops get short-circuited by the IL linker when you publish app with trimming, and we are adding more global IL linker optimizations like this in https://github.com/dotnet/runtime/pull/37615. The shared framework does not benefit from any of these optimization today that is unfortunate. I guess that this is adding more ref-counts on enabling the IL linker as post-build step to optimize the shared framework as @marek-safar suggested some time ago.

cc @eerhardt

Reply:
I would imagine that cross-gen already optimizes the shared framework in this way, does it not?

Reply:
crossgen copies IL and metadata over verbatim. crossgen will optimize what it can in the native code that it adds. The dead weight in IL and metadata will be still present in the shipping binary. 

Also, we have tiered compilation enabled by default in CoreCLR. It means that the dead weight in the IL adds work at runtime even when the shared framework is crossgened.

Reply:
I see.  It'd be an interesting experiment to measure what type of performance improvement we'd get by following type-forwards in all IL.  My pushback on running the linker at the end of the build is that it introduces a lossy step outside of developer test loop.  If the step is not lossy or is verifiably consistent with developer test loop I'm less concerned with it.

Reply:
For now let's say the following: if the project is completely type-forwards it is OK to still target System.Private.Corelib.  If a project has source we'll target System.Runtime if possible.

Reply:
We can perform those sort of optimizations as a separate PR.